source/common/http/http1/codec_impl.h

```diff
@@ -197,6 +186,91 @@ class RequestEncoderImpl : public StreamEncoderImpl, public RequestEncoder {
   bool head_request_{};
 };
 
+/**
+ * Helper class for fine grained buffer write. This class can be used to speed up some scenarios
+ * where small pieces of data are frequently written to the buffer, such as the encoding of HTTP
+ * Headers.
+ */
+class FineGrainedBufferWriteHelper {
+public:
+  static constexpr uint64_t DefaultReservationSize = 4096;
```

Comment:
I had this elsewhere but it'd be better to do this in the Buffer code rather than here, IMO.

Reply:
It's would be simple just move these code to buffer.cc/buffer.h. (The best way is to optimize `Buffer::OwnImpl` directly, but that is not easy.)

Reply:
Why is it difficult to have an API on Buffer::OwnedImpl where I can do:
```
  buffer_.write({header.name(), ": ", head.value(), "\n\n"});
```
What does the API signature look like? We could make it be `std::vector<absl::string_view>` or `std::array<absl::string_view>`, with the latter having the advantage of not requiring an extra alloc.

The impl of this can take care of ensuring there's enough contiguous space and doing the appends. You can rigorously unit test that and then use it here. I don't think this complex buffer manipulation code belongs in the h1 codec directory.

Reply:
I'm still doing some testing and thinking about a better implementation, so I haven't moved the directory of code yet.

> Why is it difficult to have an API on Buffer::OwnedImpl

Since it is not possible to implement variable-length parameter templates, we always need to construct some containers.

What I said about it not being easy was in response to your previous comment.

>  I would use some sort of templated method if possible (annoying cause it's a virtual class...) so that the same code can work whether it is supplied a C array, std::vector, std::array, or whatever, meaning at the call site you could call buffer_->appendStrings({header, ": ", value, "\r"}); without having to create allocate a temp vector. You might need a non-virtual wrapper with templated helpers.

Reply:
A var-args Buffer::write could improve performance.  I really like this idea.  It could work as a var-args method that takes string_views.  Alternatively, implement template versions that take between 1 and 10 arguments.  I doubt anyone will need more than that.

Another option would be something of a builder pattern.  Create a "buffer mutation" object that provides methods to add to the buffer using methods that bypass watermark checks, and when the mutation object goes out of scope perform all the size checks.